Importo las librerías

In [23]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Cargo los datos

In [24]:
# cargo los datos

df = pd.read_csv('C:/Users/AnaGu/Downloads/Crime_Data_from_2020_to_Present.csv')
df.info

<bound method DataFrame.info of              DR_NO               Date Rptd                DATE OCC  TIME OCC  \
0        211507896  04/11/2021 12:00:00 AM  11/07/2020 12:00:00 AM       845   
1        201516622  10/21/2020 12:00:00 AM  10/18/2020 12:00:00 AM      1845   
2        240913563  12/10/2024 12:00:00 AM  10/30/2020 12:00:00 AM      1240   
3        210704711  12/24/2020 12:00:00 AM  12/24/2020 12:00:00 AM      1310   
4        201418201  10/03/2020 12:00:00 AM  09/29/2020 12:00:00 AM      1830   
...            ...                     ...                     ...       ...   
1004984  252104112  02/02/2025 12:00:00 AM  02/02/2025 12:00:00 AM       130   
1004985  250404100  02/18/2025 12:00:00 AM  02/18/2025 12:00:00 AM      1000   
1004986  251304095  01/31/2025 12:00:00 AM  01/30/2025 12:00:00 AM      1554   
1004987  251704066  01/17/2025 12:00:00 AM  01/17/2025 12:00:00 AM      1600   
1004988  251904210  03/25/2025 12:00:00 AM  03/25/2025 12:00:00 AM      1235   

       

In [25]:
df.value_counts()

DR_NO      Date Rptd               DATE OCC                TIME OCC  AREA  AREA NAME   Rpt Dist No  Part 1-2  Crm Cd  Crm Cd Desc                                              Mocodes                                            Vict Age  Vict Sex  Vict Descent  Premis Cd  Premis Desc               Weapon Used Cd  Weapon Desc                                     Status  Status Desc   Crm Cd 1  Crm Cd 2  Crm Cd 3  Crm Cd 4  LOCATION                         Cross Street                     LAT      LON      
200613424  08/02/2020 12:00:00 AM  08/02/2020 12:00:00 AM  2030      6     Hollywood   657          1         761     BRANDISH WEAPON                                          1402 2000 1813 1814 0400 0416 0448 0329 0913 0421  34        F         H             108.0      PARKING LOT               212.0           BOTTLE                                          AO      Adult Other   761.0     920.0     930.0     998.0     WESTERN                          ROMAINE                          34.

Visualizo la tabla de crímenes

In [26]:
df

DR_NO               Date Rptd                DATE OCC  TIME OCC  \
0        211507896  04/11/2021 12:00:00 AM  11/07/2020 12:00:00 AM       845   
1        201516622  10/21/2020 12:00:00 AM  10/18/2020 12:00:00 AM      1845   
2        240913563  12/10/2024 12:00:00 AM  10/30/2020 12:00:00 AM      1240   
3        210704711  12/24/2020 12:00:00 AM  12/24/2020 12:00:00 AM      1310   
4        201418201  10/03/2020 12:00:00 AM  09/29/2020 12:00:00 AM      1830   
...            ...                     ...                     ...       ...   
1004984  252104112  02/02/2025 12:00:00 AM  02/02/2025 12:00:00 AM       130   
1004985  250404100  02/18/2025 12:00:00 AM  02/18/2025 12:00:00 AM      1000   
1004986  251304095  01/31/2025 12:00:00 AM  01/30/2025 12:00:00 AM      1554   
1004987  251704066  01/17/2025 12:00:00 AM  01/17/2025 12:00:00 AM      1600   
1004988  251904210  03/25/2025 12:00:00 AM  03/25/2025 12:00:00 AM      1235   

         AREA    AREA NAME  Rpt Dist No  Part 1-2  Crm Cd  \
0          15  N Hollywood         1502         2     354   
1          15  N Hollywood         1521         1     230   
2           9     Van Nuys          933         2     354   
3           7     Wilshire          782         1     331   
4          14      Pacific         1454         1     420   
...       ...          ...          ...       ...     ...   
1004984    21      Topanga         2103         2     946   
1004985     4   Hollenbeck          479         2     237   
1004986    13       Newton         1372         2     850   
1004987    17   Devonshire         1774         2     624   
1004988    19      Mission         1944         2     850   

                                               Crm Cd Desc  ... Status  \
0                                        THEFT OF IDENTITY  ...     IC   
1           ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT  ...     IC   
2                                        THEFT OF IDENTITY  ...     IC   
3        THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND ...  ...     IC   
4          THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)  ...     IC   
...                                                    ...  ...    ...   
1004984                          OTHER MISCELLANEOUS CRIME  ...     IC   
1004985                     CHILD NEGLECT (SEE 300 W.I.C.)  ...     IC   
1004986                                  INDECENT EXPOSURE  ...     IC   
1004987                           BATTERY - SIMPLE ASSAULT  ...     IC   
1004988                                  INDECENT EXPOSURE  ...     IC   

         Status Desc Crm Cd 1 Crm Cd 2  Crm Cd 3 Crm Cd 4  \
0        Invest Cont    354.0      NaN       NaN      NaN   
1        Invest Cont    230.0      NaN       NaN      NaN   
2        Invest Cont    354.0      NaN       NaN      NaN   
3        Invest Cont    331.0      NaN       NaN      NaN   
4        Invest Cont    420.0      NaN       NaN      NaN   
...              ...      ...      ...       ...      ...   
1004984  Invest Cont    946.0      NaN       NaN      NaN   
1004985  Invest Cont    237.0      NaN       NaN      NaN   
1004986  Invest Cont    850.0      NaN       NaN      NaN   
1004987  Invest Cont    624.0      NaN       NaN      NaN   
1004988  Invest Cont    850.0      NaN       NaN      NaN   

                                         LOCATION Cross Street      LAT  \
0         7800    BEEMAN                       AV          NaN  34.2124   
1                 ATOLL                        AV     N  GAULT  34.1993   
2        14600    SYLVAN                       ST          NaN  34.1847   
3         6000    COMEY                        AV          NaN  34.0339   
4                         4700    LA VILLA MARINA          NaN  33.9813   
...                                           ...          ...      ...   
1004984  22100    ROSCOE                       BL          NaN  34.2259   
1004985   3500    PERCY                        ST          NaN  34.0277   
1004986    300 E  53RD

In [27]:
df = df.drop(['DR_NO', 'Weapon Used Cd'], axis = 1)

Miro cuántos valores distintos tengo en las columnas y si hay muchos NaN.

In [28]:
df.value_counts()

Date Rptd               DATE OCC                TIME OCC  AREA  AREA NAME   Rpt Dist No  Part 1-2  Crm Cd  Crm Cd Desc                                              Mocodes                                            Vict Age  Vict Sex  Vict Descent  Premis Cd  Premis Desc               Weapon Desc                                     Status  Status Desc   Crm Cd 1  Crm Cd 2  Crm Cd 3  Crm Cd 4  LOCATION                         Cross Street                     LAT      LON      
01/02/2020 12:00:00 AM  01/01/2020 12:00:00 AM  2135      19    Mission     1924         1         761     BRANDISH WEAPON                                          1822 3006 3028 3039 3032 3037 3101 4019 3602       41        M         H             101.0      STREET                    KNIFE WITH BLADE 6INCHES OR LESS                AA      Adult Arrest  761.0     930.0     997.0     998.0     ASTORIA                      ST  SAN FERNANDO                 RD  34.2949  -118.4571    1
02/04/2024 12:00:00 AM  02/03/202

Podría elimianr algunas columnas que no me interesan, como 'DR_NO' que es el número de caso y el código de arma, pero mejor hago una copia del original y voy trabajando con lo que me interesa.

In [32]:
df_copy = df.copy()
df_copy = df_copy[['DATE OCC', 'TIME OCC', 'Crm Cd Desc', 'Vict Age', 'Vict Sex', 'Vict Descent', 'Premis Desc', 'Weapon Desc', 'LOCATION', 'LAT', 'LON']]
df_copy

DATE OCC  TIME OCC  \
0        11/07/2020 12:00:00 AM       845   
1        10/18/2020 12:00:00 AM      1845   
2        10/30/2020 12:00:00 AM      1240   
3        12/24/2020 12:00:00 AM      1310   
4        09/29/2020 12:00:00 AM      1830   
...                         ...       ...   
1004984  02/02/2025 12:00:00 AM       130   
1004985  02/18/2025 12:00:00 AM      1000   
1004986  01/30/2025 12:00:00 AM      1554   
1004987  01/17/2025 12:00:00 AM      1600   
1004988  03/25/2025 12:00:00 AM      1235   

                                               Crm Cd Desc  Vict Age Vict Sex  \
0                                        THEFT OF IDENTITY        31        M   
1           ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT        32        M   
2                                        THEFT OF IDENTITY        30        M   
3        THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND ...        47        F   
4          THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)        63        M   
...                                                    ...       ...      ...   
1004984                          OTHER MISCELLANEOUS CRIME        35        M   
1004985                     CHILD NEGLECT (SEE 300 W.I.C.)        11        M   
1004986                                  INDECENT EXPOSURE        16        F   
1004987                           BATTERY - SIMPLE ASSAULT        17        M   
1004988                                  INDECENT EXPOSURE        35        F   

        Vict Descent             Premis Desc  \
0                  H  SINGLE FAMILY DWELLING   
1                  H                SIDEWALK   
2                  W  SINGLE FAMILY DWELLING   
3                  A                  STREET   
4                  H                   ALLEY   
...              ...                     ...   
1004984            X                  STREET   
1004985            B  SINGLE FAMILY DWELLING   
1004986            H                  STREET   
1004987            H             HIGH SCHOOL   
1004988            H             HIGH SCHOOL   

                                            Weapon Desc  \
0                                                   NaN   
1                      KNIFE WITH BLADE 6INCHES OR LESS   
2                                                   NaN   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
1004984                                             NaN   
1004985                                             NaN   
1004986                                             NaN   
1004987  STRONG-ARM (HANDS, FIST, FEET OR BODILY FORCE)   
1004988                                             NaN   

                                         LOCATION      LAT       LON  
0         7800    BEEMAN                       AV  34.2124 -118.4092  
1                 ATOLL                        AV  34.1993 -118.4203  
2        14600    SYLVAN                       ST  34.1847 -118.4509  
3         6000    COMEY                        AV  34.0339 -118.3747  
4                         4700    LA VILLA MARINA  33.9813 -118.4350  
...                                           ...      ...       ...  
1004984  22100    ROSCOE                       BL  34.2259 -118.6126  
1004985   3500    PERCY                        ST  34.0277 -118.1979  
1004986    300 E  53RD                         ST  33.9942 -118.2701  
1004987   9600    ZELZAH                       AV  34.2450 -118.5233  
1004988  11100    OMELVENY                     AV  34.2722 -118.4417  

[1004989 rows x 11 columns]

#### Descent Code ('Vict Descent'):  
A - Other Asian  
B - Black  
C - Chinese   
D - Cambodian   
F - Filipino  
G - Guamanian  
H - Hispanic/Latin/Mexican  
I - American Indian/Alaskan Native   
J - Japanese  
K - Korean  
L - Laotian  
O - Other  
P - Pacific Islander  
S - Samoan  
U - Hawaiian  
V - Vietnamese  
W - White  
X - Unknown  
Z - Asian Indian

Voy a calcular el porcentaje de NaN por columna por si tengo que modificarlas.

In [33]:
def calcular_nan(df):
  total_filas = len(df)
  nan_por_columna = df.isnull().sum()
  porcentaje_nan = (nan_por_columna / total_filas) * 100
  return porcentaje_nan

df_nan = calcular_nan(df_copy)
print(df_nan)

DATE OCC         0.000000
TIME OCC         0.000000
Crm Cd Desc      0.000000
Vict Age         0.000000
Vict Sex        14.392595
Vict Descent    14.393789
Premis Desc      0.058508
Weapon Desc     67.437753
LOCATION         0.000000
LAT              0.000000
LON              0.000000
dtype: float64


Veo que voy a tener que eliminar la columna de 'Weapon Desc' porque hay demasiados NaN.

In [34]:
df_copy = df_copy.drop('Weapon Desc', axis = 1)
df_copy

DATE OCC  TIME OCC  \
0        11/07/2020 12:00:00 AM       845   
1        10/18/2020 12:00:00 AM      1845   
2        10/30/2020 12:00:00 AM      1240   
3        12/24/2020 12:00:00 AM      1310   
4        09/29/2020 12:00:00 AM      1830   
...                         ...       ...   
1004984  02/02/2025 12:00:00 AM       130   
1004985  02/18/2025 12:00:00 AM      1000   
1004986  01/30/2025 12:00:00 AM      1554   
1004987  01/17/2025 12:00:00 AM      1600   
1004988  03/25/2025 12:00:00 AM      1235   

                                               Crm Cd Desc  Vict Age Vict Sex  \
0                                        THEFT OF IDENTITY        31        M   
1           ASSAULT WITH DEADLY WEAPON, AGGRAVATED ASSAULT        32        M   
2                                        THEFT OF IDENTITY        30        M   
3        THEFT FROM MOTOR VEHICLE - GRAND ($950.01 AND ...        47        F   
4          THEFT FROM MOTOR VEHICLE - PETTY ($950 & UNDER)        63        M   
...                                                    ...       ...      ...   
1004984                          OTHER MISCELLANEOUS CRIME        35        M   
1004985                     CHILD NEGLECT (SEE 300 W.I.C.)        11        M   
1004986                                  INDECENT EXPOSURE        16        F   
1004987                           BATTERY - SIMPLE ASSAULT        17        M   
1004988                                  INDECENT EXPOSURE        35        F   

        Vict Descent             Premis Desc  \
0                  H  SINGLE FAMILY DWELLING   
1                  H                SIDEWALK   
2                  W  SINGLE FAMILY DWELLING   
3                  A                  STREET   
4                  H                   ALLEY   
...              ...                     ...   
1004984            X                  STREET   
1004985            B  SINGLE FAMILY DWELLING   
1004986            H                  STREET   
1004987            H             HIGH SCHOOL   
1004988            H             HIGH SCHOOL   

                                         LOCATION      LAT       LON  
0         7800    BEEMAN                       AV  34.2124 -118.4092  
1                 ATOLL                        AV  34.1993 -118.4203  
2        14600    SYLVAN                       ST  34.1847 -118.4509  
3         6000    COMEY                        AV  34.0339 -118.3747  
4                         4700    LA VILLA MARINA  33.9813 -118.4350  
...                                           ...      ...       ...  
1004984  22100    ROSCOE                       BL  34.2259 -118.6126  
1004985   3500    PERCY                        ST  34.0277 -118.1979  
1004986    300 E  53RD                         ST  33.9942 -118.2701  
1004987   9600    ZELZAH                       AV  34.2450 -118.5233  
1004988  11100    OMELVENY                     AV  34.2722 -118.4417  

[1004989 rows x 10 columns]

Miro las edades de las víctimas.

In [35]:
df_copy['Vict Age'].value_counts()

Vict Age
 0      269221
 30      22289
 35      21836
 31      21420
 29      21345
         ...  
 98         71
-2          28
-3           6
-4           3
 120         1
Name: count, Length: 104, dtype: int64

Veo que hay valores erróneos, como los negativos, 0 y el 120. Tengo que eliminarlos.  
Tengo que hacer una máscara booleana.

In [ ]:
#
#



ARREGLAR ESTO QUITANDO LOS QUE ME SOBRAN



#

mask_age = (df_copy['Vict Age'] != 0)
df_copy = df_copy[mask_age]
df_copy['Vict Age'].value_counts()

Vict Age
 30     22289
 35     21836
 31     21420
 29     21345
 28     20969
        ...  
 98        71
-2         28
-3          6
-4          3
 120        1
Name: count, Length: 103, dtype: int64

Calculo la media de esas edades.